### PR TITLE
Refactor getStatusJson

### LINKED ISSUE #270: generateExportJson follow-ups
follow-ups to #254 :

* [x] `distortableCollection.avg_cm_per_pixel`
* [x] `generateExportJson` pulls from `distortableCollection.avg_cm_per_pixel`
* [x] startExport ajax call that gets the url for the `status.json` file
* [ ] refactor to get `_getStatusJson`  separated from outer ajax call `_runExport`
* [ ] `distortableCollection.startExport(callbackFn)` with basic `console.log` callback function to show progress every X seconds

### PR DESCRIPTION
**Pull request by sashadev-sky**
Fixes #270  (<=== Add issue number here)
"refactor to get _getStatusJson separated from outer ajax call _runExport"

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updates
* [x] @mention the original creator of the issue in a comment below for help or for a review

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by sashadev-sky** [comment]
@jywarren ok I refactored it! I was thinking to merge this then write the setInterval function from the mapknitter end? It would be a mapknitter function `startExport` that calls `runExport` then `getStatusJson` on an interval

**Comment by jywarren** [comment]
I had thought the setInterval would be here, but the callback it calls would be from MapKnitter -- and we could provide a default callback here, that just, maybe, displays something super simple. What do you think?

**Comment by sashadev-sky** [comment]
@jywarren continuing the work here because its related! I can't figure out an implementation that makes sense here.

In this implementation, it allows you to call `distortableCollection.startExport()` and if you pass a cb it will call that in a `setInterval`, otherwise the default cb is `stopExport()` which just clears the interval immediately.

The issue is, if we're making the `setInterval` from inside an LDI function, then we can't return the result of `getStatusJson`, grab it in Mapknitter, and save it to the `export_url` column before calling `setInterval`. We would end up having to keep resaving it in the callback function.

**Comment by jywarren** [comment]
Ah, ok, so do we need a cb for the initial call? And a second for the
refresher?

**Comment by sashadev-sky** [comment]
@jywarren yeah would use two callbacks. Is that adding too much complexity? I will implement that if we want to go with that.

**Comment by jywarren** [comment]
No, I think that's fine, esp if both have good defaults. You're just
allowing people to override the defaults in useful ways during integration!
+1

**Comment by sashadev-sky** [comment]
@jywarren ok done & ready for merge on approval!

**Comment by jywarren** [comment]
This looks awesome. Can you add a few lines to the README to just briefly show how this can be set up, with the callbacks? Then let's merge this!

**Comment by sashadev-sky** [comment]
@jywarren For some reason it keeps updating the entire dist file here and I have to keep manually removing it but I think its fine to just let it update it? Is that ok?

Also I started getting unclear again when I was updating the documentation. 
```JS
startExport: function(collection, updater, refresher) {
    collection = collection || this.generateExportJson();
    updater = updater || false;
    refresher = refresher || this.stopExport;

    var that = this;

    $.ajax({
      url: "http://export.mapknitter.org/export",
      crossDomain: true,
      type: "POST",
      data: {
        collection: JSON.stringify(collection.images),
        scale: 30
      },
      success: function(data) {
        console.log(data);
        var exportStatus = that._getStatusJson(data);

        if (updater) { updater(exportStatus); }
        var int = setInterval(function() { refresher(exportStatus); }, 3000);

        window.int = int;
      }
    });
  },

  stopRefresher: function() {
    clearInterval(window.int);
  },

  _getStatusJson: function(data) {
    $.ajax("http://export.mapknitter.org" + data, {
      type: "GET",
      crossDomain: true
    }).done(function(exportStatus) {
      console.log(exportStatus);
      return exportStatus;
    });
  }
```
The success function in `startExport` returns a url to the `status.json` (the data parameter). So then the `exportStatus` returned after is the `status.json`. But why does this `status.json` have its own `status_url` within it leading to a different url for a `status.json`? Is this just a redirection?

So the url that we want to pass to `updater` to save is the `status_url` in that `status.json`? But to the refresher, we shouldn't actually be passing the `status.json` data because the refresher will be making a 3rd ajax call to the saved `export_url` aka a url to a completely new `status.json`?

File: README.md [hunk 1]
```diff
@@ -229,10 +229,41 @@ For multiple images, the `ToggleOrder` action switches overlapping images back a
...
 
+**`L.DistortableImageOverlay`**
+
 - [`getCorners()`](corners) and [`getCorner(idx)`](corners)
 
 - `getCenter()` - Calculates the centroid of the image
 
+**`L.DistortableCollection`**
+
+- `generateExportJson()` - for Mapknitter integration. Returns the selected images, according to the [multiple image select interface](#Multiple-Images), as a formatted JSON object. Ex:
```
Comment (jywarren) [inline README.md]:
```suggestion
- `generateExportJson()` - for full-resolution cloud-based map export, as in MapKnitter integration. Returns the selected images, according to the [multiple image select interface](#Multiple-Images), as a formatted JSON object. Ex:
```

**Comment by jywarren** [comment]
Here, I think this is right!!!

```js
startExport: function(opts) {
    opts = opts|| {};
    opts.collection = opts.collection ||  this.generateExportJson();
    opts.handleStatusUrl = opts.handleStatusUrl || _defaultHandleStatusUrl;
    opts.updater = opts.updater || _defaultUpdater;
    opts.frequency = opts.frequency || 3000;
    var statusUrl, updateInterval;

function _fetchStatusUrl(collection, handleStatusUrl)
    $.ajax({
      url: "http://export.mapknitter.org/export",
      crossDomain: true,
      type: "POST",
      data: {
        collection: JSON.stringify(collection.images),
        scale: 30
      },
      success: handleStatusUrl // this handles the initial response
    });
}

  // this may be overridden to update the UI to show export progress or completion
  function _defaultHandleStatusUrl(data) {
    statusUrl = "http://export.mapknitter.org" + data;
    $.ajax(statusUrl, {
      type: "GET",
      crossDomain: true
    }).done(function(data) {
        updateInterval = setInterval(function intervalUpdater() {
            opts.updater(data);
        }, opts.frequency);
    });
  }

  // this may be overridden to update the UI to show export progress or completion
  function _defaultUpdater(data) {
      if (data.status == "complete") clearInterval(updateInterval);
      // TODO: update to clearInterval when status == "failed" if we update that in this file:
      // https://github.com/publiclab/mapknitter-exporter/blob/main/lib/mapknitterExporter.rb
      console.log(data);
  }

_fetchStatusUrl(opts.collection, opts.handleStatusUrl);

}
```

**Comment by sashadev-sky** [comment]
@jywarren Ok I got it working but 2 things:

1) Small thing: The variable hoisting wasn't working inside the function so I had to change the order of them - is there a way to get around this or is this fine?

```JS
startExport: function(opts) {
    opts = opts|| {};
    opts.collection = opts.collection ||  this.generateExportJson();
    opts.frequency = opts.frequency || 3000;
    opts.scale = opts.scale || 30;
    var statusUrl, updateInterval;

      // this may be overridden to update the UI to show export progress or completion
    function _defaultUpdater(data) {
      if (data.status === "complete") { clearInterval(updateInterval); }
      // TODO: update to clearInterval when status == "failed" if we update that in this file:
      // https://github.com/publiclab/mapknitter-exporter/blob/main/lib/mapknitterExporter.rb
      console.log(data);
    }

    // this may be overridden to update the UI to show export progress or completion
    function _defaultHandleStatusUrl(data) {
      statusUrl = "http://export.mapknitter.org" + data;
      opts.updater = opts.updater || _defaultUpdater;

      $.ajax(statusUrl, {
        type: "GET",
        crossDomain: true
      }).done(function(data) {
          updateInterval = setInterval(function intervalUpdater() {
            opts.updater(data);
          }, opts.frequency);
      });
    }

    function _fetchStatusUrl(collection, scale) {
      opts.handleStatusUrl = opts.handleStatusUrl || _defaultHandleStatusUrl;

      $.ajax({
        url: "http://export.mapknitter.org/export",
        crossDomain: true,
        type: "POST",
        data: {
          collection: JSON.stringify(collection.images),
          scale: scale
        },
        success: opts.handleStatusUrl // this handles the initial response
      });
    }

    _fetchStatusUrl(opts.collection, opts.scale);

  }
```

2. Bigger thing:
in my pastebin I have that the example I should use as the collection to get actual status updates returned is in this link: https://pastebin.com/UK4nGCYa

But in my demo http://sasha.mapknitter.org/examples/select.html, running:
```JS
imgGroup.startExport({collection:
{"images":[{"id":53,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.51,"lng":-0.16},{"lat":51.51,"lng":-0.2},{"lat":51.49,"lng":-0.17},{"lat":51.49,"lng":-0.21}],"cm_per_pixel":36.40625},{"id":55,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.5,"lng":-0.09},{"lat":51.5,"lng":-0.13},{"lat":51.48,"lng":-0.1},{"lat":51.48,"lng":-0.14}],"cm_per_pixel":36.40625},{"id":57,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.51,"lng":-0.03},{"lat":51.51,"lng":-0.07},{"lat":51.49,"lng":-0.04},{"lat":51.49,"lng":-0.08}],"cm_per_pixel":36.40625}],"avg_cm_per_pixel":36.40625}})
```

Still returns 
```JS
{"status_url":null,"status":"starting","tms":false,"geotiff":false,"zip":false,"jpg":false,"export_id":1561602398,"user_id":null,"size":null,"width":null,"height":null,"start_time":"2019-06-27 02:26:38 +0000","run_time":null,"gem_version":"1.0.7","cm_per_pixel":null}
```
I can't recall, we found a way to fix this right? Please advise! Thank you

**Comment by sashadev-sky** [comment]
@jywarren ok for point 2 I remembered that the solution was to stop using JSON.stringify. 

So now I am getting back: 
```JS
{"status_url":"https://mapknitter-exports-warps.storage.googleapis.com/1560464327/status.json","status":"compositing","tms":false,"geotiff":"public/warps/1560464327/1560464327.tif","zip":false,"jpg":false,"export_id":1560464327,"user_id":null,"size":null,"width":null,"height":null,"start_time":"2019-06-13 22:18:46 +0000","run_time":null,"gem_version":"1.0.7","cm_per_pixel":30.0}
```

But after being called 40 times already, so thats 120 seconds, it still just says `status: compositing`. Also can't recall if there was a solution to making it actually update

**Comment by jywarren** [comment]
Hmm, if you go to
http://export.mapknitter.org/public/warps/1560464327/1560464327.tif there's
an error about public access. Assume it works for now! You can point at the
tif file for the download button. I'll follow up on the export process
itself, thanks!

**Comment by sashadev-sky** [comment]
@jywarren I got the UI working on my `gh_pages`  branch [http://sasha.mapknitter.org/examples/select.html](http://sasha.mapknitter.org/examples/select.html) and wrote a readme comment about how it works:

```
To test the multi-image interface, open `select.html`. Currently it supports multiple image selection and translations; image distortions still use the single-image interface.

  - Multiple images can be selected using <kbd>cmd</kbd> + `click` to toggle their inclusion in this interface.
  - A single toolbar instance (using `L.control`) renders the set of tools available to use on collections of images.
  - In order to return to the single-image interface, where each `L.popup` toolbar only applies actions on the image it's attached to, you must toggle *all* images out of multi-select or...
  - ...Click on the map or hit the <kbd>esc</kbd> key to quickly deselect all images.
```

But for some reason I am getting an error now on the ajax request and I can't figure out if its a rebase problem

**Comment by sashadev-sky** [comment]
If I just try to select images and run `imgGroup.startExport()` I get a jquery error

**Comment by sashadev-sky** [comment]
@jywarren Ok the interface is connected to the button now! Just need to figure out why the actual ajax requests are throwing errors and then its ready for Mapknitter. For now i'll work on making multiple image deletion along with it so that I can see through a working example
http://sasha.mapknitter.org/examples/select.html

**Comment by jywarren** [comment]
oh cool! Great work! Would a spinner icon be a good thing to show?

I see this error:

LatLng.js:69 Uncaught TypeError: Cannot read property 'lat' of undefined

**Comment by sashadev-sky** [comment]
@jywarren Do you spot anything problematic here? I don't see anything different but maybe I could have lost something in the rebase? I just have no idea if its a settings or url or code issue.

https://github.com/sashadev-sky/Leaflet.DistortableImage/blob/06d07a053bb6946d21f3ace0edf36e2c5af4dd27/src/DistortableCollection.js#L274-L351

Spinner is a good plan! I can use the `setInterval` and a `clearInterval` for now just to get the UI functioning if we fix the error being thrown!

**Comment by jywarren** [comment]
Looking now, thanks!

**Comment by jywarren** [comment]
Ok, from this error:

```

  | toString | @ | LatLng.js:69
-- | -- | -- | --
  | add | @ | jquery.js:8685
  | buildParams | @ | jquery.js:8672
  | buildParams | @ | jquery.js:8666
  | (anonymous) | @ | jquery.js:8653
  | each | @ | jquery.js:367
  | buildParams | @ | jquery.js:8644
  | buildParams | @ | jquery.js:8666
  | (anonymous) | @ | jquery.js:8653
  | each | @ | jquery.js:367
  | buildParams | @ | jquery.js:8644
  | jQuery.param | @ | jquery.js:8709
  | ajax | @ | jquery.js:9301
  | _fetchStatusUrl | @ | DistortableCollection.js:337
  | startExport | @ | DistortableCollection.js:349
  | addHooks | @ | DistortableImage.EditToolbar2.js:29
  | enable | @ | leaflet.toolbar.js:1
  | enable | @ | leaflet.toolbar.js:1
  | handler | @ | DomEvent.js:79
```

I think jQuery is having trouble turning `collection.images` into JSON. But it's engaging some Leaflet code. Maybe the Leaflet and jQuery aren't playing nice? It's thinking an `undefined` is a `LatLng` object, and trying to stringify it:

```
	// @method toString(): String
	// Returns a string representation of the point (for debugging purposes).
	toString: function (precision) {
		return 'LatLng(' +
		        Util.formatNum(this.lat, precision) + ', ' +
```

Um, Leaflet can parse `[0,0]`-style 2-item arrays into `LatLng` objects, so maybe it's mistaking one of our coordinates for one of these, somehow?

**Comment by jywarren** [comment]
OK, when i run `imgGroup.generateExportJson()`, i get:
```
{"images":[{"id":143,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.5,"lng":-0.13},{"lat":51.5,"lng":-0.09},{"lat":51.48,"lng":-0.14},{"lat":51.48,"lng":-0.1}],"cm_per_pixel":36.40625},{"id":176,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.51,"lng":-0.07},{"lat":51.51,"lng":-0.03},{"lat":51.49,"lng":-0.08},{"lat":51.49,"lng":-0.04}],"cm_per_pixel":36.40625}],"avg_cm_per_pixel":36.40625}
```

So, no big clue there...

**Comment by jywarren** [comment]
OK, interestingly, the lat/lng tuples are actually `LatLng` objects, probably because of how we're assembling the JSON: 

image

Maybe we need to make them plain objects?

**Comment by jywarren** [comment]
Can we change this line to just create a plain object?

https://github.com/sashadev-sky/Leaflet.DistortableImage/blob/06d07a053bb6946d21f3ace0edf36e2c5af4dd27/src/DistortableCollection.js#L290

**Comment by jywarren** [comment]
Yeah so maybe this is what's going wrong:

```
i.images[1].nodes[0].toString()
> "LatLng(51.51, -0.07)"
```

**Comment by jywarren** [comment]
OK, so I managed to replace the `LatLng` objects with pure JSON, but then got this error from the export service:

```
NoMethodError at /export
undefined method `[]' for nil:NilClass
file: app.rb location: run_export line: 95
BACKTRACE
(expand)

JUMP TO: GET POST COOKIES ENV
app.rb in run_export
  scale = params[:scale] || images_json[0]['cm_per_pixel']
```

**Comment by jywarren** [comment]
OK, so you really do need to just `JSON.stringify` the entire JSON collection youre sending up, or it will send `LatLng` objects. I think that's the first error. Then, you're getting a CORS error because you're missing something in the JSON, related to scale. Let's see.

**Comment by jywarren** [comment]
OK, got it. We need to be uploading /just/ the array, not the enclosing JSON object brackets. So instead of:

```js
{"images":
[{"id":127,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.5,"lng":-0.13},{"lat":51.5,"lng":-0.09},{"lat":51.48,"lng":-0.14},{"lat":51.48,"lng":-0.1}],"cm_per_pixel":36.40625},{"id":160,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.51,"lng":-0.07},{"lat":51.51,"lng":-0.03},{"lat":51.49,"lng":-0.08},{"lat":51.49,"lng":-0.04}],"cm_per_pixel":36.40625}]
}
```

We need just:

```js
[{"id":127,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.5,"lng":-0.13},{"lat":51.5,"lng":-0.09},{"lat":51.48,"lng":-0.14},{"lat":51.48,"lng":-0.1}],"cm_per_pixel":36.40625},{"id":160,"src":"http://sasha.mapknitter.org/examples/example.png","nodes":[{"lat":51.51,"lng":-0.07},{"lat":51.51,"lng":-0.03},{"lat":51.49,"lng":-0.08},{"lat":51.49,"lng":-0.04}],"cm_per_pixel":36.40625}]
```

Then it works!

**Comment by jywarren** [comment]
Also, if at all possible, let's stop using `window.scale`, too!

Can you upload your latest code here?

**Comment by jywarren** [comment]
Just curious, why do we need to fill the image group with things on initialize, if they're not selected yet? 

https://github.com/sashadev-sky/Leaflet.DistortableImage/blob/336640a4125f2493cc1c097de3997fca7b2794a0/examples/select.html#L101-L109

**Comment by sashadev-sky** [comment]
@jywarren is it ok to open a new PR and push the code there? there is some code here like documentation that is not on my `gh-pages` branch and rebasing will get complicated.

1. `window.scale` was a straggler, removed now!
2. I had thought that `collection: JSON.stringify(collection.images)` is not going to work, that is what was giving me `status_url: null` and changing it to `collection.images` was meant to be the solution. So if I put this back to JSON.stringify, it does in fact stop giving me the latlng error but now i'll just get `status_url: none` again. Which I think is a worse problem right?

I am going to try having a new method that will take in `getCorners()` and convert it into a plain JS obj instead of a latLng instance like you suggested the first time around! Not sure if that'll help but worth a try? 

3. We can fill the image group with things whenever we'd like, it is dynamic now. But the main diff is that, before we were adding images to the map and feature group, but really the feature group is meant to be used without adding them to the map too. This was making event and UI handling convoluted. We cannot call `img.editing.enable` until the images are added to the map through the group, though. 

   - The group implementation has not changed, it has always known about every image on the map. It manages their group state (multi-selected or not). Individual images still track their own individual selected state as well (via `this._selected`), but if group selection is on it gets precedence.

4. Today I will be trying to extend the API to allow for manipulation of toolbar actions. I need: 1) a way to add and remove our own tools. This would be enough to make it work with MK now. Even better would be 2) the actions to be accessible so that we can optionally pass our `startExport` callbacks to the `Exports` tool

**Comment by sashadev-sky** [comment]
going to look at @rexagod PR and see if I can grab anything from there. https://github.com/publiclab/Leaflet.DistortableImage/pull/155. Is there a way I can commit credit him if I use it? Rebasing it at this point is going to be very hard

**Comment by jywarren** [comment]
I tested with stringify back in, and it worked! I wonder if the other
status_url: null was because of the extra { images: } in our JSON POST?

I wouldn't mess with the LatLng - the more testing I did, the more I found
if we add Stringify then the JSON submission is fine just except for the
extra { images: } wrapping the array.

**Comment by jywarren** [comment]
Great, thanks! Sure, you can mention him in the commit?

**Comment by sashadev-sky** [comment]
@jywarren new PR is at #314!

**Comment by jywarren** [comment]
Completed in #314!

**Pull request closed without merging**